make models more accessible in python

NippunSharma opened the pull request:
This is in reference to #1709.
I have added `SerializeOutJSON` function that outputs the model to a json using `cereal`. Also I have added a utility function `process_params` that takes in the model and outputs the parameters.
So the overall procedure would look like this:

    >>> from mlpack import linear_rergession
    >>> from mlpack.preprocess_json_params import process_params
    >>> X = [1,2,3,4,5,6,7,8,9,10]
    >>> y = [1,2,3,4,5,6,7,8,9,10]
    >>> output = linear_regression(training=X, training_responses=y)
    >>> model = output["output_model"]
    >>> params_str, params_dic = process_params(model, True)
 
here `params_str` is a formatted string and printing the string would give something like this:

    >>> print(params_str)
    {
        "LinearRegression": {
            "cereal_class_version": 0,
            "parameters": {
                "n_rows": "2",
                "n_cols": "1",
                "vec_state": "1",
                "elem": 0.0,
                "elem": 1.0
            },
            "lambda": 0.0,
            "intercept": True
        }
    }

and `params_dic` is the same string converted to a python dictionary.

Things to add:
I feel that maybe there is some more post processing required on the JSON but was not able to decide exactly what to do. So I have made this PR and maybe we can discuss the post processing here.

- [x] convert arma elements to numpy vectors
- [x] remove `cereal_class_version`

rcurtin commented:
Very nice!  Cereal makes this *way* easier than it would have been with boost::serialization.

At the very least, we should probably process Armadillo elements into numpy vectors... so for instance,

```
        "parameters": {
            "n_rows": "2",
            "n_cols": "1",
            "vec_state": "1",
            "elem": 0.0,
            "elem": 1.0
        },
```

should become a numpy array of type `float64`, with elements `[0.0 1.0]`.  (Note that I transposed that---in Armadillo it's a column vector.  But numpy makes things "look" row major, so when going back and forth between numpy and Armadillo we do an implicit transpose.)

It would be really cool if we could allow the user to modify the JSON representation, then be able to ingest it again (so, our preprocessing would be reversible), but that could be a bunch of extra work.  Even just making these parameters visible to the user is an awesome start. :+1:

Let me know what you think!  There are probably some other nicenesses we can do, like stripping `cereal_class_version` perhaps.

NippunSharma commented:
Hey @rcurtin, I have added the preprocessing in the latest commit. It can be used as follows:

```python
from mlpack import linear_regression
from mlpack.preprocess_json_params import process_params

X = [1,2,3,4]
y = [1,2,3,4]

out = linear_regression(training=X, training_responses=y)
model = out["output_model"]
params_dic, params_str = process_params(model, True, True)
```
I have added the function that removes `cereal_class_version`, and a function that converts armadillo to numpy array.
I have also added an argument to `pretty print` the dictionary.

I am testing it with different methods in mlpack, I will get back to you with the results.

NippunSharma commented:
Outputs for different mlpack methods:

`Regression`:

    {'LinearRegression': {'intercept': True,
                          'lambda': 0.0,
                          'parameters': array([[0.5, 0.1]])}}
    
    {'BayesianLinearRegression': {'alpha': 16.264802428447386,
                                  'beta': 3.461553590036266,
                                  'centerData': False,
                                  'dataOffset': array([], shape=(1, 0), dtype=float64),
                                  'dataScale': array([], shape=(1, 0), dtype=float64),
                                  'gamma': 0.8645970377976727,
                                  'matCovariance': array([[0.0083256]]),
                                  'maxIterations': 50,
                                  'omega': array([[0.23055921]]),
                                  'responsesOffset': 0.0,
                                  'scaleData': False,
                                  'tolerance': 0.0001}}
    
    {'LARS': {'activeSet': [0],
              'betaPath': [{'7486': 0.0,
                            'n_cols': '1',
                            'n_rows': '1',
                            'vec_state': '1'},
                           {'8474': 0.2666666666666666,
                            'n_cols': '1',
                            'n_rows': '1',
                            'vec_state': '1'}],
              'elasticNet': False,
              'ignoreSet': [],
              'isActive': [True],
              'isIgnored': [False],
              'lambda1': 0.0,
              'lambda2': 0.0,
              'lambdaPath': [8.0, 1.7763568394002505e-15],
              'lasso': False,
              'matGramInternal': array([[30.]]),
              'matUtriCholFactor': array([], shape=(0, 0), dtype=float64),
              'tolerance': 1e-16,
              'useCholesky': False}}

`Classification`:

    {'DecisionTreeModel': {'info': {'maps': [], 'types': [False]},
                           'tree': {'classProbabilities': array([[0.75, 0.25]]),
                                    'dimensionTypeOrMajorityClass': 1,
                                    'splitDimension': 448630121828,
                                    'value0': {'vecSize': 0}}}}
    
    {'AdaBoostModel': {'dimensionality': 1,
                       'mappings': array([[1., 0.]]),
                       'value0': {'smartPointer': {'ptr_wrapper': {'data': {'alpha': [0.5493061443340549,
                                                                                      0.0],
                                                                            'numClasses': 2,
                                                                            'tolerance': 1e-10,
                                                                            'wl': [{'classProbabilities': array([[0.75, 0.25]]),
                                                                                    'dimensionTypeOrMajorityClass': 0,
                                                                                    'splitDimension': 0,
                                                                                    'value0': {'vecSize': 0}},
                                                                                   {'classProbabilities': array([[0.5, 0.5]]),
                                                                                    'dimensionTypeOrMajorityClass': 1,
                                                                                    'splitDimension': 0,
                                                                                    'value0': {'vecSize': 0}}]},
                                                                   'valid': 1}}},
                       'weakLearnerType': 0}}
    
    {'LinearSVMModel': {'mappings': array([[0., 1.]]),
                        'svm': {'fitIntercept': True,
                                'lambda': 0.0001,
                                'numClasses': 2,
                                'parameters': array([[-0.15882892,  0.15881829],
           [ 0.02367202, -0.02365268]])}}}
    
    {'LogisticRegression[]': {'lambda': 0.0,
                              'parameters': array([[-124.24117089],
           [  49.46242369]])}}
    
    {'PerceptronModel': {'map': array([[1., 0.]]),
                         'p': {'biases': array([[-3.,  3.]]),
                               'maxIterations': 1000,
                               'weights': array([[-1.],
           [ 1.]])}}}
    
    {'NBCModel': {'mappings': array([[0., 1.]]),
                  'nbc': {'means': array([[3.66666667],
           [2.        ]]),
                          'probabilities': array([[0.75, 0.25]]),
                          'variances': array([[3.33333333e-01],
           [1.00000000e-10]])}}}
    
    {'SoftmaxRegression': {'fitIntercept': True,
                           'lambda': 0.0001,
                           'numClasses': 2,
                           'parameters': array([[ 10.85344471,  -4.37673559],
           [-10.85347363,   4.37657496]])}}

@rcurtin  Can you take a look at these outputs and let me know what more things we should process here?
I have not shown outputs for `hoeffding_tree` and `random_forest` because they were very big and difficult to interpret.

rcurtin commented:
@NippunSharma that looks *excellent*!  Of course, the models with tree structures are going to be too difficult for a user to look at, but they could theoretically parse through them programmatically.

Let me know if you're ready for me to review it.  I'm excited to get this in---visibility into models will be super useful for users.  As an aside, we don't need to do it in this PR, but is the processing reversible?  Could I, e.g., get a JSON representation of a model like you've shown here, play with some parameters, then process it back into C++?  It seems like it should be conceptually possible, although it might be a bit more work.  We can open an issue and handle it later, of course.  Even what you've shown here is a huge step forwards.

NippunSharma commented:
Hey @rcurtin, thanks! 
I have also added the reversible processing now. I have added the function `feed_params` that sets the model parameters using the given dictionary.
I guess this is ready for review. 

I have tested the reversed procedure for `linear_regression` and it seems to be working.
The reversed procedure will look like this:

```python
from mlpack import linear_regression
import numpy as np
from mlpack.preprocess_json_params import process_params, feed_params

X = [1,2,3,4]
y = [1,2,3,4]
X_test = [5,6]

out = linear_regression(training=X, training_responses=y)
model = out["output_model"]

preds_before = linear_regression(input_model=model, test=X_test)
params_dic = process_params(model)
print("Before...")
print(params_dic)
print(preds_before["output_predictions"],"\n")

# doubling the parameters, so the predictions will also be doubled now.
params_dic["LinearRegression"]["parameters"] = np.array([[0.,2.]])
feed_params(model, params_dic)

preds_after = linear_regression(input_model=model, test=X_test)
print("After...")
print(params_dic)
print(preds_after["output_predictions"])
```
Output:
```bash
Before...
{'LinearRegression': {'cereal_class_version': 0, 'parameters': array([[0., 1.]]), 'lambda': 0.0, 'intercept': True}}
[5. 6.] 

After...
{'LinearRegression': {'cereal_class_version': 0, 'parameters': array([[0., 2.]]), 'lambda': 0.0, 'intercept': True}}
[10. 12.]
```

NippunSharma commented:
In the above example, I am changing the parameters using a python dictionary, this can also be changed by writing parameters to a JSON file and then reading in that JSON file. What do you think is a better option? ...we can also have both options.

NippunSharma commented:
@rcurtin, we should probably add some documentation for this feature also, what do you think? Also, should there be some unit tests that test the reversible procedure?

NippunSharma commented:
Hey @rcurtin, can you please take a look at this whenever you get some time.

rcurtin reviewed (commented):
Strange, it looks like I started this review a few weeks ago but never actually posted it!  Anyway, hopefully my comments here are helpful. :+1: I think overall it is fairly close if we can address the comments.  Let me know if I can clarify anything! :)

NippunSharma commented on src/mlpack/bindings/python/mlpack/preprocess_json_params.py:
```diff
@@ -0,0 +1,184 @@
+    for i in range(len(obj)):
+      np_to_arma(obj[i], rand_gen)
+  else:
+    pass
```
this is a recursive function. here we will `pass` whenever we encounter an object that is not a dictionary or a list because then we cannot recurse further.

rcurtin replied:
Oh, ok, I overlooked that!  Thanks for the clarification. :+1:

NippunSharma commented:
Hey @rcurtin, thanks for the reviews.
Sorry that this took some time, I have made quite a lot of changes here. As it turns out, creating the correct json that can be ingested back into cereal is a rather tough task --
1) `cereal_class_version` should be present in the ingested json.
It seemed easy to accomplish the task at the beginning but there was a big obstacle in the way. Each json output by cereal had more than one occurrence of `cereal_class_version` and there was no visible pattern to these occurrences. So, while removing `cereal_class_version` I had to store the complete path of each occurrence so that we can insert them back at the correct places.
2) the order of keys should remain the same.
This one was the hardest to debug. As it turns out, the order of keys matters while ingesting a json into cereal, it must be the same as the json that cereal gave us. To tackle this issue, I am using `OrderedDict` instead of a regular python dictionary as it preserves the order of keys and provides some nice functions using which we can manipulate the order easily.

Apart from these I also changed the way we're handling the `same keys` problem here. I have used a more pythonic and standard way of handling this situation now rather than just replacing each occurrence with a number.

Also, I felt that it might get confusing for the user to use the utility functions that we have provided. So, I decided that a better approach would be to just do the preprocessing behind the scenes and just give the users a dictionary and take from the user a dictionary. All preprocessing is done automatically by mlpack.
The interface now looks like this: 
```python
from mlpack import decision_tree
X = [1,2,3,4]
y = [1,0,1,1]
out = decision_tree(training=X, labels=y)
model = out["output_model"]
params = model.get_cpp_params()
# do some changes to the params.
# ...
# now set the params back into the model.
model.set_cpp_params(params)
```

changing the name of methods from "get_param" to "get_cpp_param" and "set_param" to "set_cpp_param" is to make the user aware that these parameters are actually coming from c++ implementation.

I will add the documentation for this functionality in the coming few days.

Let me know what you think of the changes that I made here?

rcurtin commented on src/mlpack/bindings/python/mlpack/preprocess_json_params.py:
```diff
@@ -0,0 +1,267 @@
+      insert_in_dic(params_dic_copy, path, param_name, val)
+
+  # dumping to string. restore_value defined later.
+  params_str = json.dumps(params_dic_copy,  cls=restore_value)
```
```suggestion
  params_str = json.dumps(params_dic_copy, cls=restore_value)
```
Tiny spacing fix. :)

rcurtin commented on src/mlpack/bindings/python/mlpack/preprocess_json_params.py:
```diff
@@ -0,0 +1,267 @@
+  def encode(self, o):
+    if isinstance(o, dict):
+      if "elem" in o.keys():
+        to_return = '{%s' % ', '.join(': '.join((json.encoder.py_encode_basestring(k), self.encode(v))) for k, v in o.items() if k != "elem")
```
Would you mind wrapping lines to 80 characters to match the style guide? :)

rcurtin commented on src/mlpack/bindings/python/mlpack/preprocess_json_params.py:
```diff
@@ -0,0 +1,267 @@
+  after following a particular path.
+  '''
+  temp = dic[path[0]]
+  for idx in range(1,len(path)):
```
```suggestion
  for idx in range(1, len(path)):
```
Tiny style fix :)

rcurtin reviewed (commented):
> As it turns out, creating the correct json that can be ingested back into cereal is a rather tough task

Wow, you are right!  That is way more complicated than I would have expected.  But, awesome that you got it figured out!

> changing the name of methods from "get_param" to "get_cpp_param" and "set_param" to "set_cpp_param" is to make the user aware that these parameters are actually coming from c++ implementation.

I agree, I think those are clearer names.

> I will add the documentation for this functionality in the coming few days.

Sounds good!

> Let me know what you think of the changes that I made here?

I took a look through and everything looks great to me!  I left a few minor comments.

I think overall once the documentation is added, this one is ready to merge.  Thank you so much for working on this!  I had originally imagined that this would be a huge undertaking and be a ton of code, but as it turns out, your implementation here is pretty concise and straightforward. :+1:

jeffin143 commented:
May be add a line to history too , this is a nice addition :) good job

rcurtin commented on src/mlpack/bindings/python/print_type_doc_impl.hpp:
```diff
@@ -155,9 +155,13 @@ std::string PrintTypeDoc(
-      "in C++ and inspected there.";
+      "model.  This model pointer has 2 methods using which the parameters "
+      "of the model can be inspected as well as changed through Python.  "
+      "The get_cpp_params() method returns a python ordered dictionary that "
```
```suggestion
      "The `get_cpp_params()` method returns a python ordered dictionary that "
```

rcurtin commented on src/mlpack/bindings/python/print_type_doc_impl.hpp:
```diff
@@ -155,9 +155,13 @@ std::string PrintTypeDoc(
-      "model.  Note that this means that the mlpack model itself cannot be "
-      "easily inspected in Python; however, the pickled model can be loaded "
-      "in C++ and inspected there.";
+      "model.  This model pointer has 2 methods using which the parameters "
```
```suggestion
      "model.  This model pointer has 2 methods with which the parameters "
```
Just a simple grammar fix. :)

rcurtin commented on src/mlpack/bindings/python/print_type_doc_impl.hpp:
```diff
@@ -155,9 +155,13 @@ std::string PrintTypeDoc(
+      "contains all the parameters of the model.  The user can inspect the "
+      "parameters as well change the parameter values in the dictionary "
+      "(without deleting any keys) and pass that back into the model "
+      "using the set_cpp_params() method.";
```
```suggestion
```
Oops, my last suggestion forgot to include this line...

rcurtin commented on src/mlpack/bindings/python/print_type_doc_impl.hpp:
```diff
@@ -155,9 +155,13 @@ std::string PrintTypeDoc(
+      "model.  This model pointer has 2 methods using which the parameters "
+      "of the model can be inspected as well as changed through Python.  "
+      "The get_cpp_params() method returns a python ordered dictionary that "
+      "contains all the parameters of the model.  The user can inspect the "
+      "parameters as well change the parameter values in the dictionary "
+      "(without deleting any keys) and pass that back into the model "
```
```suggestion
      "contains all the parameters of the model.  These parameters can "
      "be inspected and changed.  To set new parameters for a model, "
      "pass the modified dictionary (without deleting any keys) to the "
      "`set_cpp_params()` method."
```
I think maybe we can clarify this part a little bit---let me know what you think. :+1:

(I also didn't check closely that everything here is 80 characters, so it's possible it could be wrapped better.)

rcurtin reviewed (commented):
Hey @NippunSharma, looks great!  Just some final suggestions on the documentation.  Let me know what you think. :+1:

rcurtin reviewed (approved):
Awesome work @NippunSharma!  This is a big step forward in usability for the Python bindings. :rocket:
